docs/proposals/ldap-group-sync.md

```diff
@@ -0,0 +1,105 @@
+# Syncing LDAP Groups to OpenShift Groups
+
+Linking OpenShift `Groups` to their representations on LDAP servers allows for administrators to have an easier way to manage `Groups` inside OpenShift. This additional functionality will require no additional fields on the OpenShift `Group` object. The sync job will indicate that an OpenShift `Group` mirrors an external representation by setting the `openshift.io/externalRecordProvider` `Annotation` in the `Group` as well as adding a `Label` such as `openshift.io/externalRecordProvider/LDAP` to the `Group`. Other information (e.g. last sync time or resource version) will be stored as `Annotations` on the OpenShift `Group`. Logistical information for the LDAP sync will be stored in an `LDAPSyncOptions` that the administrator will need to provide.
+
+### Use Cases
+1. I want to sync all LDAP groups I've configured in a sync job (e.g. a specific LDAP server configuration) to OpenShift `Groups`
+* I want to sync specific OpenShift `Groups` with their `externalRecordProviders`
+* I want to sync a speficic set of LDAP groups to OpenShift `Groups`
+* I want to run a destructive sync that will remove OpenShift groups that were created or updated by a previous sync if the LDAP server no longer contains their record
+* I want to map an LDAP group to an existing OpenShift `Group`
+
+## Supported LDAP Group Schemas
+The three LDAP schemas supported are
+
+1. groups as first-class LDAP entries that contain lists of members of the group
+2. users as first-class LDAP entries that contain lists of groups they are a member of
+3. schema 2 above, also including first-class groups to store metadata
+
+### Data Flow
+* Determine list of LDAP groups to be synced
+* Find LDAP group members
+* Determine OpenShift `Identities` from LDAP group members
+* Determine OpenShift `Users` from OpenShift `Identities`
+* Determine OpenShift `Group` from LDAP group
+* Populate new OpenShift `Group`'s `Users` from resulting OpenShift `UserIdentityMappings`, update OpenShift `Group` metadata fields tied to LDAP attributes, leave other OpenShift `Group` fields unchanged
+
+##  Determining Which LDAP Groups To Sync
+An `LDAPGroupLister` determines what LDAP groups needs to be synced and outputs the result as a set of unique identifier strings. The other objects that take these unique identifiers need to understand the format of this string (e.g. these objects will be tightly coupled). For example, an `LDAPGroupLister` for schema 1 above cannot be used with a `LDAPGroupDataExtractor` for schema 2. The four `LDAPGroupLister` implementations that are necessary are:
```

Comment:
"as a set of unique identifier strings, called LDAP groupUIDs" so we can reasonable refer to them in the rest of the doc.